### PR TITLE
Generate Constructor should not copy "protected" from abstract base class (#25238)

### PR DESCRIPTION
**Pull request by angelina-dev**
fixes : https://github.com/dotnet/roslyn/issues/25238

There was no answer to my question in comments for the issue: "If a base class is not abstract, it looks like we should make the constructor of the derived class public. What do you think?". So the implementation did not affect this case and a constructor is still generated with the same access modifier as base constructor.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/GenerateDefaultConstructors/GenerateConstructorForAbstractClass.cs [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Microsoft.CodeAnalysis.GenerateDefaultConstructors;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.GenerateDefaultConstructors
+{
+    public class GenerateConstructorForAbstractClassTests : AbstractCSharpCodeActionTest
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/GenerateDefaultConstructors/GenerateConstructorForAbstractClass.cs]: Thanks for the PR :)

we don't need a new test class for these tests.  It should just go in the primary test class htat already exists for this feature.  You can put it in another partial-part if you want to (though i would not recommend that personally).

File: src/EditorFeatures/CSharpTest/GenerateDefaultConstructors/GenerateConstructorForAbstractClass.cs [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    {
+        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
+            => new GenerateDefaultConstructorsCodeRefactoringProvider();
+
+        [WorkItem(25238, "https://github.com/dotnet/roslyn/issues/25238")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateDefaultConstructors)]
+        public async Task TestGenerateConstructorFromProtectedConstructor()
+        {
+            await TestInRegularAndScriptAsync(
+@"abstract class C : [||]B
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/GenerateDefaultConstructors/GenerateConstructorForAbstractClass.cs]: you don't have to change up all these test.  But i found them odd to read as you had to read out of order.  isntead of seeing hte base type, then the dependent type, you see the dependent type and then have to find the base type.

File: src/Features/Core/Portable/GenerateMember/GenerateDefaultConstructors/AbstractGenerateDefaultConstructorsService.AbstractCodeAction.cs [hunk 3]
```diff
@@ -50,19 +50,23 @@ private abstract class AbstractCodeAction : CodeAction
...
-                var baseConstructorArguments = constructor.Parameters.Length != 0
-                    ? syntaxFactory.CreateArguments(constructor.Parameters)
+                var baseConstructorArguments = baseConstructor.Parameters.Length != 0
+                    ? syntaxFactory.CreateArguments(baseConstructor.Parameters)
                     : default;
 
+                var classType = _state.ClassType;
+                var accessibility = baseConstructor.ContainingType.IsAbstractClass() ? 
+                    (classType.IsAbstractClass() ? baseConstructor.DeclaredAccessibility : Accessibility.Public)
+                    : baseConstructor.DeclaredAccessibility;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/GenerateMember/GenerateDefaultConstructors/AbstractGenerateDefaultConstructorsService.AbstractCodeAction.cs]: IDE style would be:

```c#
var accessibility = baseConstructor.ContainingType.IsAbstractClass()
    ? classType.IsAbstractClass() ? baseConstructor.DeclaredAccessibility : Accessibility.Public
    : baseConstructor.DeclaredAccessibility;
```

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/GenerateMember/GenerateDefaultConstructors/AbstractGenerateDefaultConstructorsService.AbstractCodeAction.cs]: or:

```c#
var accessibility = baseConstructor.ContainingType.IsAbstractClass()
    ? classType.IsAbstractClass()
        ? baseConstructor.DeclaredAccessibility
        : Accessibility.Public
    : baseConstructor.DeclaredAccessibility;

Comment (Neme12) [reply src/Features/Core/Portable/GenerateMember/GenerateDefaultConstructors/AbstractGenerateDefaultConstructorsService.AbstractCodeAction.cs]: interesting bit of logic... I would personally just prefer:
```c#
var accessibility  = baseConstructor.ContainingType.IsAbstractClass() && !classType.IsAbstractClass()
    ? Accessibility.Public
    : baseConstructor.DeclaredAccessibility;
```

File: src/EditorFeatures/CSharpTest/GenerateDefaultConstructors/GenerateConstructorForAbstractClass.cs (hunk 1 above)
Comment (angelina-dev) [reply src/EditorFeatures/CSharpTest/GenerateDefaultConstructors/GenerateConstructorForAbstractClass.cs]: Thank you for Code Review :+1: 
Done.

File: src/EditorFeatures/CSharpTest/GenerateDefaultConstructors/GenerateConstructorForAbstractClass.cs (hunk 2 above)
Comment (angelina-dev) [reply src/EditorFeatures/CSharpTest/GenerateDefaultConstructors/GenerateConstructorForAbstractClass.cs]: Yeah, I agree. The tests were written in the same way as in GenerateDefaultConstructorsTests.cs.

File: src/Features/Core/Portable/GenerateMember/GenerateDefaultConstructors/AbstractGenerateDefaultConstructorsService.AbstractCodeAction.cs (hunk 3 above)
Comment (angelina-dev) [reply src/Features/Core/Portable/GenerateMember/GenerateDefaultConstructors/AbstractGenerateDefaultConstructorsService.AbstractCodeAction.cs]: Done like in Neme12's suggestion

**Comment by heejaechang** [comment]
retest windows_release_unit32_prtest  please

**Comment by heejaechang** [comment]
@jinujoseph 15.7?

**Comment by jinujoseph** [comment]
@angelina-dev  Thanks much for the PR.
@heejaechang , Lets take this in 15.8 , don't see a reason to rush for 15.7 (correct me if there is one)

**Comment by heejaechang** [comment]
@jinujoseph for 15.8 do I need approval? or sign off is all I need?

**Comment by jinujoseph** [comment]
no ask approvals needed 
review sign off + test pass needed

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure jenkins issue

D:\j\workspace\windows_relea---a23004dd\Binaries\Tools\dotnet\sdk\2.1.300-preview2-008324\NuGet.targets(114,5): error : Failed to retrieve information about 'Microsoft.VisualStudio.Text.UI.Wpf' from remote source

https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_release_unit32_prtest/12806/console

**Comment by heejaechang** [comment]
retest windows_release_unit32_prtest please

**Comment by heejaechang** [comment]
again, Jenkins issue. myget is down.

**Comment by heejaechang** [comment]
retest windows_release_unit32_prtest please

**Comment by DustinCampbell** [comment]
Awesome! I've been wanting this for awhile.

**Pull request merged**